utils/notification_manager.py

```diff
@@ -77,22 +80,35 @@ async def new_deposit(new_crypto_balances: dict, deposit_amount_usd, telegram_id
         await NotificationManager.send_to_admins(message, user_button, bot)
 
     @staticmethod
-    async def new_buy(subcategory_id: int, quantity: int, total_price: float, user: User, bot):
-        subcategory = await SubcategoryService.get_by_primary_key(subcategory_id)
-        message = ""
+    #async def new_buy(subcategory_id: int, quantity: int, total_price: float, user: User, bot):
+    async def new_buy(sold_cart_items: List[CartItem], user: User, bot):
+
         username = user.telegram_username
         telegram_id = user.telegram_id
         user_button = await NotificationManager.make_user_button(username)
-        if username:
-            message += Localizator.get_text_from_key("new_purchase_notification_with_tgid").format(
-                username=username,
-                total_price=total_price,
-                quantity=quantity,
-                subcategory_name=subcategory.name)
-        else:
-            message += Localizator.get_text_from_key("new_purchase_notification_with_username").format(
-                telegram_id=telegram_id,
-                total_price=total_price,
-                quantity=quantity,
-                subcategory_name=subcategory.name)
+        cart_grand_total = 0.0
+        message = ""
+        line_max_width = 0
+
+        for cart_item in sold_cart_items:
+            subcategory = await SubcategoryService.get_by_primary_key(cart_item.subcategory_id)
+            line_max_width = max (line_max_width, len(str(subcategory)))
+            cart_item_total = cart_item.a_piece_price * cart_item.quantity
+            cart_grand_total += cart_item_total
+
+            if username:
+                message += Localizator.get_text_from_key("new_purchase_notification_with_tgid").format(
+                    username=username,
+                    total_price=cart_item_total,
+                    quantity=cart_item.quantity,
+                    subcategory_name=subcategory.name) + "\n"
+            else:
+                message += Localizator.get_text_from_key("new_purchase_notification_with_username").format(
+                    telegram_id=telegram_id,
+                    total_price=cart_item_total,
+                    quantity=cart_item.quantity,
+                    subcategory_name=subcategory.name) + "\n"
+        dashes = "-" * min(config.MAX_LINE_WITH, line_max_width)
+        message += dashes + "\n"
+        message += Localizator.get_text_from_key("cart_grand_total_string").format(cart_grand_total=cart_grand_total, cart_item_currency=config.CURRENCY)
```

Comment:
To be honest, I don't really like those dashes.
You can use Telegram styles to make the final message more beautiful.

https://core.telegram.org/bots/api#html-style

<b>bold</b>, <strong>bold</strong>
<i>italic</i>, <em>italic</em>
<u>underline</u>, <ins>underline</ins>
<s>strikethrough</s>, <strike>strikethrough</strike>, <del>strikethrough</del>
<span class="tg-spoiler">spoiler</span>, <tg-spoiler>spoiler</tg-spoiler>
<b>bold <i>italic bold <s>italic bold strikethrough <span class="tg-spoiler">italic bold strikethrough spoiler</span></s> <u>underline italic bold</u></i> bold</b>
<a href="http://www.example.com/">inline URL</a>
<a href="tg://user?id=123456789">inline mention of a user</a>
<tg-emoji emoji-id="5368324170671202286">👍</tg-emoji>
<code>inline fixed-width code</code>
<pre>pre-formatted fixed-width code block</pre>
<pre><code class="language-python">pre-formatted fixed-width code block written in the Python programming language</code></pre>
<blockquote>Block quotation started\nBlock quotation continued\nThe last line of the block quotation</blockquote>
<blockquote expandable>Expandable block quotation started\nExpandable block quotation continued\nExpandable block quotation continued\nHidden by default part of the block quotation started\nExpandable block quotation continued\nThe last line of the block quotation</blockquote>